### PR TITLE
SQL-785: Database property is required

### PR DESCRIPTION
**Pull request by nbagnard**

### CODE REVIEW

File: src/main/java/com/mongodb/jdbc/MongoDriver.java [hunk 1]
```diff
@@ -299,11 +318,13 @@ private static String removePrefix(String prefix, String s) {
 
     // getConnectionString constructs a valid MongoDB connection string which will be used as an input to the mongoClient.
     // If there are required fields missing, those fields will be returned in DriverPropertyInfo[] with a null connectionString
-    private Pair<ConnectionString, DriverPropertyInfo[]> getConnectionString(
+    private Pair<ConnectionString, DriverPropertyInfo[]> getConnectionSettings(
             String url, Properties info) throws SQLException {
+        List<DriverPropertyInfo> propertyInfoList = new ArrayList<>();
```
Comment (bucaojit) [inline src/main/java/com/mongodb/jdbc/MongoDriver.java:323]: I don't see the `propertyInfoList` variable used?

File: src/main/java/com/mongodb/jdbc/MongoDriver.java [hunk 2]
```diff
@@ -153,39 +151,59 @@ protected MongoConnection getUnvalidatedConnection(String url, Properties info)
...
-                throw new SQLException("password specified without user");
-            }
-            if (driverPropertyInfo[0].name.equals(PASSWORD)) {
-                throw new SQLException("user specified without password");
-            }
-            String[] propertyNames = new String[driverPropertyInfo.length];
-            for (int i = 0; i < propertyNames.length; ++i) {
-                propertyNames[i] = driverPropertyInfo[i].name;
+        if (missingRequiredProperties.length != 0) {
+            for (DriverPropertyInfo info : missingRequiredProperties) {
```
Comment (bucaojit) [inline src/main/java/com/mongodb/jdbc/MongoDriver.java:173]: Not sure I understand the logic here.  There is a for loop through the properties, however it looks like the first one will always throw the  SQLException.  Or was it meant to loop through the properties and append the errors to a string and throw that exception at the end?

File: src/main/java/com/mongodb/jdbc/MongoDriver.java [hunk 3]
```diff
@@ -221,25 +239,26 @@ private MongoConnection createConnection(ConnectionString cs, Properties info)
...
+                                    + " is "
+                                    + MONGOSQL_DIALECT);
                 }
-                return new MongoSQLConnection(cs, info.getProperty(DATABASE), logLevel, logDir);
+                return new MongoSQLConnection(cs, database, logLevel, logDir);
             default:
-                throw new SQLClientInfoException(
-                        String.format("invalid dialect '%s'", dialect),
-                        Collections.singletonMap(DIALECT, ClientInfoStatus.REASON_VALUE_INVALID));
+                throw new SQLException(
```
Comment (bucaojit) [inline src/main/java/com/mongodb/jdbc/MongoDriver.java:255]: Good catch, it looks like `SQLClientInfoException` is mainly thrown in the `setClientInfo()` method.

File: src/test/java/com/mongodb/jdbc/MongoDriverTest.java [hunk 4]
```diff
@@ -87,113 +90,112 @@ void testDBURL() throws SQLException {
...
+            fail("The connection should fail because a mandatory connection setting is missing.");
+        } catch (SQLException e) {
+            // Expected failure
+            assertEquals(
+                    MongoDriver.CONNECTION_ERROR_SQLSTATE,
+                    e.getSQLState(),
+                    "Expect SQL state "
+                            + MongoDriver.CONNECTION_ERROR_SQLSTATE
+                            + " but got "
+                            + e.getMessage());
```
Comment (bucaojit) [inline src/test/java/com/mongodb/jdbc/MongoDriverTest.java:107]:
```suggestion
                            + e.getSQLState());
```
Suggestion to match the type, SQL state.

**Review by bucaojit — CHANGES_REQUESTED** [review]

File: src/test/java/com/mongodb/jdbc/MongoDriverTest.java (hunk 4 above)
Comment (nbagnard) [reply src/test/java/com/mongodb/jdbc/MongoDriverTest.java]: You are right!

**Review by pmeredit — APPROVED** [review]
LGTM

**Comment by nbagnard** [comment]
evergreen merge

**Pull request merged**